src/wfs/src/main/java/org/geoserver/wfs/Transaction.java

```diff
@@ -620,181 +629,106 @@ public void dataStoreChange(TransactionEvent event) throws WFSException {
-                    deleteElementHandler.execute(
-                            deleteElement, request, stores, result, multiplexer);
-                    deleteElement.deleteFilter();
+            if (aggrTargetElement instanceof Delete && pElem instanceof Delete) {
+                if (aggrDeleteCount == maxDeleteCount) {
+                    return false;
+                }
+                Delete lTarget = (Delete) aggrTargetElement;
+                Delete lElem = (Delete) pElem;
+                QName lTargetType = lTarget.getTypeName();
+                QName lElemType = lElem.getTypeName();
+                if (lTargetType != null && lTargetType.equals(lElemType)) {
+                    return true;
                 }
-            } else {
-                deleteElement = element;
-            }
-            return deleteElement;
-        }
-
-        private Insert executeBulkAndGetInsert(
-                Insert insertElement,
-                Delete deleteElement,
-                TransactionElementHandler deleteElementHandler,
-                Insert element) {
-            // since batching is only enabled for INSERT and DELETE, push batches of DELETE
-            // operations
-            if (batchType != BatchType.INSERT) {
-                executeDeleteBulkTransaction(deleteElement, deleteElementHandler);
             }
-            if (insertElement != null) {
-                Insert insert = element;
-                insertElement.addFeature(insert.getFeatures());
-            } else {
-                insertElement = element;
+            return false;
+        }
+
+        private void aggregate(TransactionElement pElem) {
+            boolean lRemoveFromRequest = false;
+            if (aggrTargetElement instanceof Insert) {
+                Insert lTarget = (Insert) aggrTargetElement;
+                Insert lElem = (Insert) pElem;
+                lTarget.addFeatures(lElem.getFeatures());
+                lRemoveFromRequest = true;
+            } else if (aggrTargetElement instanceof Delete) {
+                Delete lTarget = (Delete) aggrTargetElement;
+                Delete lElem = (Delete) pElem;
+                lTarget.addFilter(lElem.getFilter());
+                aggrDeleteCount++;
+                lRemoveFromRequest = true;
             }
-            return insertElement;
-        }
-
-        // execute batch inserts then clean the batch
-        private void executeInsertBulkTransaction(
-                Insert insertElement, TransactionElementHandler insertElementHandler) {
-            if (insertElement != null && batchType == BatchType.INSERT) {
-                insertElementHandler.execute(insertElement, request, stores, result, multiplexer);
-                insertElement.setFeatures(new ArrayList());
-                batchType = BatchType.DELETE;
+            if (lRemoveFromRequest) {
+                // contents of the element have been added to target element. To avoid contents
```

Comment:
Fair point, but the current execution model does not spare them from inconsistency anyways.
When runAggregated() is run, the listeners receive a transaction request object (part of TransactionEvent), which is being modified as you go, so each call returns a different state of it.

At the same time, DispatcherCallback can pick the parsed request object either at the leading or trailing edge of execution, and they would get two different views of the request if it gets modified.

If you want to avoid all inconsistencies, then a new TransactionRequest object would have to be built, completely restructured, and then executed at the end (giving a consistent vision to the transaction listeners) while preserving a consistent vision to the dispatcher callbacks.

Reply:
I understand your point, but I would not call a modified or restructured request object *inconsistent* per se.

What I meant with "inconsistent" is that contents logically differ - unintentionally. E.g. previously batch manager *copied* features from the Transactions Insert #2 to Insert #1, *leaving* them in Insert #1 also. 

The current implementation just *moves* the features. But from a logical perspective the request object stays the same.

From my point of view it is not required that the request object stays unmodified for all listeners, etc. It is sufficient if it is logically the same.

Ok?

Reply:
Fair enough.